portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -16,61 +16,85 @@
 
 import java.io.IOException;
 import java.util.*;
-import com.google.appengine.api.datastore.DatastoreService;
-import com.google.appengine.api.datastore.DatastoreServiceFactory;
-import com.google.appengine.api.datastore.Entity;
-import com.google.appengine.api.datastore.PreparedQuery;
-import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.*;
 import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import javax.servlet.RequestDispatcher;
 import com.google.gson.*;
 
-/** Servlet that returns some example content. */
+/** Servlet that returns comment data created by the user. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int numComments = 0;
+
+    try {
+        numComments = Integer.parseInt(request.getParameter("number"));
+    } catch (NumberFormatException e) {
+        // Leave the default value alone.
+    }
+    
     Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
-    int numComments = Integer.parseInt(request.getParameter("number"));
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    ArrayList<String> comments = new ArrayList<String>();
+    /* If no comments are returned by the query, set a descriptive HTTP status code to signify an error. */
+    int size = results.countEntities(FetchOptions.Builder.withLimit(1000));
```

Comment:
Since you're calling `results.countEntities` and `results.asIterable` separately, you're actually hitting the database twice. Both calls also don't use the `maxComments` limit, so they're fetching a bunch of data only to discard it.

You can optimize this by getting the iterator up front, passing the limit as a parameter, and checking for the presence of any comments with `hasNext`:

```java
Iterable<Entity> commentEntities =
    results.asIterable(FetchOptions.limit(maxComments));
if (!commentEntities.iterator().hasNext()) {
  response.sendError(404);
  return;
}

// `commentEntities` is now guaranteed to have no more than `maxComments`
// comments, so we can just use a `foreach` loop:
for (Entity commentEntity : commentEntities) {
  // ...
```